oak_restricted_kernel/src/syscall/dice_data.rs

```diff
@@ -16,42 +16,119 @@
 
-struct DiceDataDescriptor {
+pub enum DiceData {
+    #[cfg(feature = "initrd")]
+    Layer0(Box<Stage0DiceData>),
+    Layer1(Box<RestrictedKernelDiceData>),
+}
+
+impl DiceData {
+    fn as_mut_slice(&mut self) -> &mut [u8] {
+        match self {
+            #[cfg(feature = "initrd")]
+            DiceData::Layer0(stage0_dice_data) => {
+                <Stage0DiceData as zerocopy::AsBytes>::as_bytes_mut(stage0_dice_data)
+            }
+            DiceData::Layer1(stage1_dice_data) => {
+                <RestrictedKernelDiceData as zerocopy::AsBytes>::as_bytes_mut(stage1_dice_data)
+            }
+        }
+    }
+}
+
+struct ReadState {
     data: DiceData,
     index: usize,
 }
 
+struct WriteState {
+    data: RestrictedKernelDiceData,
+    index: usize,
+}
+
+enum DiceDataDescriptor {
+    Readable(Box<ReadState>),
+    #[cfg(feature = "initrd")]
+    Writeable(Box<WriteState>),
+}
+
 impl DiceDataDescriptor {
     fn new(data: DiceData) -> Self {
-        Self { index: 0, data }
+        Self::Readable(Box::new(ReadState { index: 0, data }))
     }
 }
 
 impl FileDescriptor for DiceDataDescriptor {
     fn read(&mut self, buf: &mut [u8]) -> Result<isize, oak_restricted_kernel_interface::Errno> {
-        let data_as_slice = <DiceData as zerocopy::AsBytes>::as_bytes_mut(&mut self.data);
-        let length = copy_max_slice(&data_as_slice[self.index..], buf);
+        match self {
+            #[cfg(feature = "initrd")]
+            DiceDataDescriptor::Writeable(_write_state) => Err(Errno::EINVAL),
+            DiceDataDescriptor::Readable(read_state) => {
+                let data_as_slice = read_state.data.as_mut_slice();
+                let length = copy_max_slice(&data_as_slice[read_state.index..], buf);
 
-        // destroy the data that was read, to ensure that it can only be read once
-        let slice_to_read = &mut data_as_slice[self.index..(self.index + length)];
-        slice_to_read.zeroize();
+                // destroy the data that was read, to ensure that it can only be read once
+                let slice_to_read =
+                    &mut data_as_slice[read_state.index..(read_state.index + length)];
+                slice_to_read.zeroize();
 
-        self.index += length;
-        Ok(length as isize)
+                read_state.index += length;
+                Ok(length as isize)
+            }
+        }
     }
 
-    fn write(&mut self, _buf: &[u8]) -> Result<isize, oak_restricted_kernel_interface::Errno> {
-        // Writing is not supported.
-        Err(Errno::EINVAL)
+    fn write(&mut self, buf: &[u8]) -> Result<isize, oak_restricted_kernel_interface::Errno> {
+        match self {
+            DiceDataDescriptor::Readable(read_state) => match &mut read_state.data {
+                #[cfg(feature = "initrd")]
+                DiceData::Layer0(stage0_dice_data) => {
+                    <Stage0DiceData as zerocopy::AsBytes>::as_bytes_mut(stage0_dice_data).zeroize();
```

Comment:
alternatively we could only allow writing after the data has been fully read (and in the process destroyed). But imo this is safer, if less efficient.